### PR TITLE
feat: add compose file katana and madara

### PR DESCRIPTION
**Pull request by ftupas**
Time spent on this PR: 0.5

Depends On: [Deployer Image](https://github.com/kkrt-labs/kakarot/pkgs/container/kakarot%2Fdeployer) pushed ✅
Resolves: #318 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

Hello, this pr does the following:
- build and push to docker in parallel for different platforms
- add compose file to run Katana, deploy Kakarot to it, and initialize the Kakarot RPC server
- add compose file to run Madara, deploy Kakarot to it, and initialize the Kakarot RPC server

Sample hurl call to RPC (Katana)
image

Sample hurl call to RPC (Madara)
image

# Does this introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: crates/core/src/client/config.rs
Comment (ftupas) [inline crates/core/src/client/config.rs:38]: Make the RPC URL configurable to be able to connect to containerized instances

File: crates/core/src/client/constants.rs
Comment (ftupas) [inline crates/core/src/client/constants.rs:18]: Move default examples to `.env.examples`

File: docker-compose.katana.yaml
Comment (ClementWalter) [inline docker-compose.katana.yaml:6]: We usually run katana with these

```
--validate-max-steps 16777216 --invoke-max-steps 16777216 --gas-price 0
```

Comment (ftupas) [reply docker-compose.katana.yaml:6]: Gotcha let me update this with those values 🙂

File: docker-compose.katana.yaml
Comment (jobez) [inline docker-compose.katana.yaml:44]: we need the deployer process to fund an eoa and usually set the private key to match with the first test account of anvil, so setting
`EVM_PRIVATE_KEY=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80` right here would help make the forge script easily runnable!

File: docker-compose.madara.yaml
Comment (jobez) [inline docker-compose.madara.yaml:23]: here as well: 
`EVM_PRIVATE_KEY=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80`

**Review by jobez — DISMISSED** [review]
thanks so much for your help fred!

File: docker-compose.katana.yaml
Comment (ftupas) [reply docker-compose.katana.yaml:44]: Made `EVM_PRIVATE_KEY` configurable and the one from `.env`

**Review by ftupas — COMMENTED** [review]
Thanks for the reviews! @ClementWalter @jobez  🙏

File: crates/core/src/client/config.rs
Comment (Eikix) [reply crates/core/src/client/config.rs:38]: Can we avoid unwraps in this case?

File: crates/core/src/client/constants.rs
Comment (Eikix) [reply crates/core/src/client/constants.rs:18]: Nice

**Review by ClementWalter — CHANGES_REQUESTED** [review]
I think that this PR can be simplified. The deployer image doesn't need ot use network specific env variables.

How I see things:
- a main `docker-compose.yml` file with a `starknet` service running the madara image
- `kakarot-deployer` is used with plain `RPC_URL`, `ACCOUNT_ADDRESS` and `PRIVATE_KEY`
- deployments are in `deployments/deployment.json` (no network name)
- run the kakarot-rpc without using the "predefined" network katana/madara etc but just passing the RPC_URL and KAKAROT_ADDRESS env var (not sure there is anything to change in the rpc repo actually)

then we can also add a `docker-compose.katana.yml` that just redefine the `starknet` service to use the katana image, and run

```
docker compose up -f docker-compose.yml -f docker-compose.katana.yml
```

File: docker-compose.katana.yaml
Comment (ftupas) [inline docker-compose.katana.yaml:4]: This overrides the `starknet` service and uses `katana` for the devnet

File: docker-compose.katana.yaml
Comment (ftupas) [inline docker-compose.katana.yaml:21]: Hardcore these values instead of using `.env` to reduce error surface

File: docker-compose.yaml
Comment (ftupas) [inline docker-compose.yaml:5]: Use `madara` as the default devnet

File: docker-compose.yaml
Comment (ftupas) [inline docker-compose.yaml:23]: Hardcore these values instead of using .env to reduce error surface

**Review by ftupas — COMMENTED** [review]
Thanks for suggestions @ClementWalter ! I've made the docker flow simpler and removed the reliance on the `.env` file. Also reverted the changes on the rpc repo and use the custom RPC URL functionality 👍

File: README.md
Comment (ClementWalter) [inline README.md:161]: you should enable markdownlint in your ide!

File: README.md
Comment (ClementWalter) [inline README.md:175]: I don't think the `compute-units-per-second` is relevant here, see https://docs.alchemy.com/reference/compute-units

File: docker-compose.yaml
Comment (ClementWalter) [inline docker-compose.yaml:52]: declare also a `volume` to avod dumping deployments in local host

File: docker-compose.yaml
Comment (ClementWalter) [inline docker-compose.yaml:44]: I don't really lmike this `run.sh` for just exporting 2 env variable from the previous run, though I'm not sure how to simply remove it

**Review by ClementWalter — CHANGES_REQUESTED** [review]
I left some comments in the compose file for adding a `volume`

File: docker-compose.yaml
Comment (ClementWalter) [reply docker-compose.yaml:44]: Actually we could do what we did for the devnet version, see https://github.com/kkrt-labs/kakarot/blob/a0b5112004e7ec573a30ac5fc31ec8093f6cb07d/.github/workflows/deployments.yml#L50

**Review by ftupas — COMMENTED** [review]
Thanks for the review @ClementWalter !

File: README.md
Comment (ClementWalter) [inline README.md:156]: still no markdownlint in your IDE I guess ! (note: using only vscode from github I review the PR from, I see MD013/line-length: Line length [Expected: 80; Actual: 140]markdownlintMD013
)

File: docker-compose.yaml
Comment (ClementWalter) [inline docker-compose.yaml:47]: `tini` is meant to be the entrypoint of the image, see [tini doc](https://github.com/krallin/tini)

Following on our previous comment regardig $GITHUB_OUTPUT, it think that a legit step could be using the `alpine` image to convert the deployments into a .env file used in `kakarot-rpc: env_file` section, or just to overwrite the `kakarot-deployer image to create such a file at the end (I don't like this one because it makes our file break if the deployer changes, but not such a big deal I think)

**Review by ClementWalter — COMMENTED** [review]
made another suggestion for using the artifacts!

File: README.md
Comment (ftupas) [reply README.md]: Managed to fix this and so I also did a fix on other parts of the README

File: docker-compose.yaml
Comment (ftupas) [reply docker-compose.yaml]: Saw this on the [tini doc](https://github.com/krallin/tini#existing-entrypoint) about existing entrypoints and tried it out and it works for our usecase 🤓 let me know what you think

image

File: README.md
Comment (ftupas) [inline README.md:216]: Couldn't fix this one so I added an ignore for this portion

File: docker-compose.yaml
Comment (ftupas) [inline docker-compose.yaml:47]: We override the entrypoint but still use `tini` to override it as per [docs](https://github.com/krallin/tini#existing-entrypoint)

File: README.md
Comment (ClementWalter) [reply README.md:216]: anyway I'll clean the README once this is merged, this info is not accurate anymore!

File: docker-compose.yaml
Comment (ClementWalter) [inline docker-compose.yaml:35]: can we use https://hub.docker.com/r/apteno/alpine-jq/tags instead, it's still maintained, open source and has clean tags

File: docker-compose.yaml
Comment (ClementWalter) [inline docker-compose.yaml:38]: Can we do with this instead?

```
entrypoint: /bin/sh -c
command: |
    echo ...
    echo ...
```

it splits better the image part and what we do with it

File: docker-compose.yaml
Comment (ftupas) [reply docker-compose.yaml]: Gotcha, thanks for the find! Using this instead 👍

**Review by ClementWalter — DISMISSED** [review]
approved! but if you feel so, you could run `katana` to the same address:port as madara and remove some of the katana.yaml overriding

**Comment by ClementWalter** [comment]
See the katana help:

```
  -p, --port <PORT>
          Port number to listen on.

          [default: 5050]

      --host <HOST>
```

**Review by ClementWalter — APPROVED** [review]

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**
